File: backend/mini-job-service/src/main/java/com/example/minijob/job/service/JobService.java
// [FILE]
// - 목적: Job 비즈니스 로직 처리
// - 주요 역할: Job 생성, 조회, 상태 변경 등의 서비스 레이어 로직
// - 관련 토이 버전: [BE-v0.2], [BE-v0.3]+
// - 권장 읽는 순서: 생성자 → createJob() → getJob() → getAllJobs() → deleteJob()
//
// [LEARN] 서비스 레이어는 비즈니스 로직을 담당한다.
//         컨트롤러와 저장소 사이에서 트랜잭션 경계를 관리한다.

package com.example.minijob.job.service;

import com.example.minijob.job.domain.Job;
import com.example.minijob.job.repository.JobRepository;
import com.example.minijob.common.exception.JobNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

// [Order 1] Job 서비스
// - @Service로 스프링 빈 등록
// - @Transactional로 트랜잭션 관리
// - 토이 버전: [BE-v0.2]+
// [LEARN] 서비스 레이어에서 비즈니스 규칙을 적용하고,
//         여러 저장소를 조합한 복잡한 작업을 처리한다.
@Service
@Transactional(readOnly = true)
public class JobService {

    private final JobRepository jobRepository;

    public JobService(JobRepository jobRepository) {
        this.jobRepository = jobRepository;
    }

    // [Order 2] Job 생성
    // - 새 Job을 생성하고 저장
    // - 토이 버전: [BE-v0.2]+
    // [LEARN] @Transactional을 메서드에 붙이면 쓰기 작업이 가능해진다.
    //         클래스 레벨의 readOnly=true를 오버라이드한다.
    @Transactional
    public Job createJob(String type, String payload) {
        Job job = new Job(type, payload);
        return jobRepository.save(job);
    }

    // [Order 3] ID로 Job 조회
    // - 없으면 예외 발생
    // [LEARN] Optional을 사용하여 null 처리를 명시적으로 한다.
    public Job getJob(Long id) {
        return jobRepository.findById(id)
                .orElseThrow(() -> new JobNotFoundException("Job not found with id: " + id));
    }

    // [Order 4] 전체 Job 목록 조회
    public List<Job> getAllJobs() {
        return jobRepository.findAll();
    }

    // [Order 5] 상태별 Job 조회
    public List<Job> getJobsByStatus(String status) {
        return jobRepository.findByStatus(status);
    }

    // [Order 6] Job 삭제
    // - 토이 버전: [BE-v0.2]+
    @Transactional
    public void deleteJob(Long id) {
        if (!jobRepository.existsById(id)) {
            throw new JobNotFoundException("Job not found with id: " + id);
        }
        jobRepository.deleteById(id);
    }

    // [Order 7] Job 상태 변경
    // - 토이 버전: [BE-v0.7]에서 스케줄러가 사용
    @Transactional
    public Job updateJobStatus(Long id, String status) {
        Job job = getJob(id);
        job.setStatus(status);
        return jobRepository.save(job);
    }
}

File: backend/mini-job-service-node/src/job/jobService.ts
// [FILE]
// - 목적: Job 비즈니스 로직 집약
// - 주요 역할: 검증, 조회, 상태 변경, 스케줄러 훅 제공
// - 관련 토이 버전: [Node-BE-v0.1]
// - 권장 읽기 순서: 생성자 → CRUD → 헬퍼 → 스케줄 시뮬레이션
//
// [LEARN] 이 서비스 레이어는 컨트롤러/라우터에서 받은 요청을 검증하고 도메인 객체를 조작한다.
//         각 메서드는 명확한 역할을 가지고 있으며, 상태 전이는 도메인 객체의 메서드를 통해 수행된다.
import { HttpError } from '../common/errorHandler';
import { Job, JobStatus } from './job';
import { JobRepository } from './jobRepository';

export class JobService {
  constructor(private readonly repo: JobRepository) {}

  // [Order 1] Job 생성
  // 요청에서 전달받은 type과 payload를 검증하고 저장소에 위임하여 새 Job을 반환합니다.
  // - 입력 검증: type이 비어있으면 400 에러
  // - 성공 시: 생성된 Job 객체 반환
  createJob(type: string, payload?: string): Job {
    if (!type) {
      throw new HttpError(400, 'type is required');
    }
    return this.repo.save(type, payload);
  }

  // [Order 2] Job 목록 조회
  // 저장소에 저장된 모든 Job을 ID 순으로 반환합니다.
  listJobs(): Job[] {
    return this.repo.findAll();
  }

  // [Order 3] 단일 Job 조회
  // 주어진 ID로 저장소에서 Job을 찾습니다. 존재하지 않으면 404를 던집니다.
  getJob(id: string): Job {
    if (!id) {
      throw new HttpError(400, 'id is required');
    }

    const job = this.repo.findById(id);
    if (!job) {
      throw new HttpError(404, `Job ${id} not found`);
    }
    return job;
  }

  // [Order 4] Job 삭제
  // ID로 Job을 삭제합니다. 존재하지 않으면 404를 반환합니다.
  deleteJob(id: string): void {
    const removed = this.repo.delete(id);
    if (!removed) {
      throw new HttpError(404, `Job ${id} not found`);
    }
  }

  // [Order 4-1] Job 취소 (실패 상태로 표시)
  // - 비동기 작업 취소를 흉내 내기 위해 FAILED 상태를 사용한다.
  cancelJob(id: string): Job {
    const job = this.getJob(id);
    job.fail();
    return job;
  }

  // [Order 5] Job 상태 전이
  // 주어진 상태로 Job을 변경합니다. PENDING은 생성 시 자동으로 설정됩니다.
  // - IN_PROGRESS: start() 호출
  // - COMPLETED: complete() 호출
  // - FAILED: fail() 호출
  transition(id: string, status: JobStatus): Job {
    const job = this.getJob(id);
    if (status === 'IN_PROGRESS') job.start();
    if (status === 'COMPLETED') job.complete();
    if (status === 'FAILED') job.fail();
    return job;
  }

  // [Order 6] 다음 처리 대기 Job 찾기
  // scheduler가 호출하여 가장 먼저 생성된 PENDING Job을 찾습니다. 없으면 undefined.
  findNextPending(): Job | undefined {
    return this.listJobs().find((job) => job.status === 'PENDING');
  }
}